Adding the functionality for specify packaged project filename and title

SeqLaz opened the pull request:
This is complement for the [PR](https://github.com/opengisch/qfieldsync/pull/615) on QFieldSync

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -85,6 +85,8 @@ def __init__(
         self,
         project: QgsProject,
         export_folder: str,
+        export_filename: str,
+        export_title: str,
```
```suggestion
        export_filename: str,
        export_title: str,
```

These should be optional because we want to be backward compatible with QFieldCloud.

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -103,6 +105,8 @@ def __init__(
         self.trUtf8 = self.tr
 
         self.export_folder = Path(export_folder)
+        self._export_filename = Path(export_filename).stem
```
```suggestion
        self.export_folder = Path(export_filename).parent
        self._export_filename = Path(export_filename).stem
```

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -135,7 +139,16 @@ def convert(self, reload_original_project: bool = True) -> None:
         project = QgsProject.instance()
         self.original_filename = Path(project.fileName())
-        self.backup_filename = make_temp_qgis_file(project)
+        if (
+            self.export_type == ExportType.Cable
+            and self._export_filename
+            and self._export_title
+        ):
+            self.backup_filename = make_temp_qgis_file(
+                project, self._export_filename, self._export_title
+            )
+        else:
+            self.backup_filename = make_temp_qgis_file(project)
```
```suggestion
        self.backup_filename = make_temp_qgis_file(
            project, self._export_filename, self._export_title
        )
```

This should work the same no matter cable or not.

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -272,9 +285,14 @@ def _convert(self, project: QgsProject) -> None:
-        export_project_filename = self.export_folder.joinpath(
-            f"{self.original_filename.stem}_qfield.qgs"
-        )
+        if self.export_type == ExportType.Cable and self._export_filename:
+            export_project_filename = self.export_folder.joinpath(
+                f"{Path(self.backup_filename).stem}_qfield.qgs"
+            )
+        else:
+            export_project_filename = self.export_folder.joinpath(
+                f"{self.original_filename.stem}_qfield.qgs"
+            )
```
```suggestion
        export_project_filename = self._export_filename
```

I guess in QFieldSync you need to make sure this file does not exists (yet) and ask for confirmation to overwrite it.

suricactus commented on libqfieldsync/utils/qgis.py:
```diff
@@ -48,12 +48,22 @@ def open_project(filename: str, filename_to_read: Optional[str] = None) -> bool:
-    backup_filename = os.path.join(project_backup_dir, f"{project.baseName()}.qgs")
-    project.write(backup_filename)
-    project.setFileName(original_filename)
+    if exported_filename and exported_project_title:
+        backup_filename = os.path.join(project_backup_dir, f"{exported_filename}.qgs")
+        project.write(backup_filename)
+        project.setTitle(exported_project_title)
+        project.setFileName(backup_filename)
+    else:
+        backup_filename = os.path.join(project_backup_dir, f"{project.baseName()}.qgs")
+        project.write(backup_filename)
+        project.setFileName(original_filename)
```
```suggestion
    project.write(exported_filename)
    project.setFileName(exported_filename)

    if exported_project_title:
        project.setTitle(exported_project_title)
```

I guess something like this.

suricactus reviewed (commented):
We did interactive review with @SeqLaz

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -135,7 +138,9 @@ def convert(self, reload_original_project: bool = True) -> None:
         project = QgsProject.instance()
         self.original_filename = Path(project.fileName())
-        self.backup_filename = make_temp_qgis_file(project)
+        self.backup_filename = make_temp_qgis_file(
+            project, self._export_filename, self._export_title
+        )
```
Why you need this? Isn't the export actually done as part of line 280 in the `export_project_filename` variable?

SeqLaz replied:
Now I notice!.... I will create the respective changes!

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -102,7 +103,9 @@ def __init__(
-        self.export_folder = Path(export_folder)
+        self.export_folder = Path(export_filename).parent
+        self._export_filename = Path(export_filename).stem
+        self._export_title = export_title
```
You need to pass this to `convert_to_offline()` method of `BaseOffliner` instances.

SeqLaz replied:
Noted!

suricactus reviewed (changes requested):
Still not there.

If you accept the comments from the PR, please accept them in batch instead of individual commits.

Also note that the CI is failing. We need to fix this too, don't overlook it.

Finally, we should also add a test about this.

SeqLaz commented:
> Still not there.
> 
> If you accept the comments from the PR, please accept them in batch instead of individual commits.
> 
> Also note that the CI is failing. We need to fix this too, don't overlook it.
> 
> Finally, we should also add a test about this.

I squashed all the commits into only one commit; after all, there was no need for many changes!

suricactus commented on libqfieldsync/utils/qgis.py:
```diff
@@ -48,13 +48,19 @@ def open_project(filename: str, filename_to_read: Optional[str] = None) -> bool:
 
 
-def make_temp_qgis_file(project: QgsProject) -> str:
+def make_temp_qgis_file(
+    project: QgsProject,
+    exported_project_title: Optional[str] = None,
+) -> str:
```
I don't think this change is needed, since you pass the title to the offliners.

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -317,9 +320,7 @@ def _convert(self, project: QgsProject) -> None:
-                gpkg_filename,
-                offline_layers,
-                bbox,
+                gpkg_filename, offline_layers, bbox, self._export_title
```
```suggestion
                gpkg_filename,
                offline_layers,
                bbox,
                self._export_title,
```

We prefer trailing comma on long lists to reduce the diff size.

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -135,7 +138,7 @@ def convert(self, reload_original_project: bool = True) -> None:
         project = QgsProject.instance()
         self.original_filename = Path(project.fileName())
-        self.backup_filename = make_temp_qgis_file(project)
+        self.backup_filename = make_temp_qgis_file(project, self._export_title)
```
```suggestion
        self.backup_filename = make_temp_qgis_file(project)
```

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -102,7 +103,9 @@ def __init__(
         self.trUtf8 = self.tr
 
-        self.export_folder = Path(export_folder)
+        self.export_folder = Path(export_filename).parent
+        self._export_filename = Path(export_filename).stem
```
```suggestion
        self._export_filename = Path(export_filename)
```

Can we explore if we can kill the `export_folder` attribute and use `self._export_filename.parent` instead?

Note that `_export_filename` has changed from the `stem` to the full filename. If we keep the previous state, we need to rename it to `_export_basename`.

suricactus commented on libqfieldsync/offliners.py:
```diff
@@ -144,8 +146,11 @@ def convert_to_offline(
     ) -> bool:
-        self._convert_to_offline_project(str(offline_db_filename), layers, bbox)
+        self._convert_to_offline_project(
+            str(offline_db_filename), layers, bbox, exported_project_title
```
```suggestion
            str(offline_db_filename),
            layers,
            bbox,
            exported_project_title,
```

suricactus commented on libqfieldsync/offliners.py:
```diff
@@ -53,6 +53,7 @@ def convert_to_offline(
         offline_db_filename: str,
         layers: List[QgsMapLayer],
         bbox: Optional[QgsRectangle],
+        exported_project_title: Optional[str] = None,
```
```suggestion
        exported_project_title: str = "",
```
I think this will siplify the API around, no?

suricactus commented on libqfieldsync/offliners.py:
```diff
@@ -86,6 +87,7 @@ def convert_to_offline(
         offline_db_filename: str,
         layers: List[QgsMapLayer],
         bbox: Optional[QgsRectangle],
+        exported_project_title: Optional[str] = None,
```
```suggestion
        exported_project_title: Optional[str] = None,
```

Make sure you use this parameter here.

suricactus reviewed (changes requested):
Getting there, please add the title thing to the `QgisCoreOffliner`.

SeqLaz commented:
> Getting there, please add the title thing to the `QgisCoreOffliner`.

Hey @suricactus  please have a look.
I added the option to specify the project title passed to the offliner `QfisCoreOffliner`. And addressed all the recommendations.

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -266,14 +268,16 @@ def _convert(self, project: QgsProject) -> None:
 
-        export_project_filename = self.export_folder.joinpath(
-            f"{self.original_filename.stem}_qfield.qgs"
+        export_project_filename = self._export_filename.parent.joinpath(
+            f"{self._export_filename.stem}.qgs"
         )
```
```suggestion
        export_project_filename = self._export_filename
```

This is enough, we can add a check earlier that `self._export_filename` is a `.qgs` file.

This should also be handled in the UI.

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -84,14 +84,15 @@ class OfflineConverter(QObject):
         export_type: ExportType = ExportType.Cable,
         create_basemap: bool = True,
         dirs_to_copy: Optional[Dict[str, bool]] = None,
+        export_title: Optional[str] = None,
```
```suggestion
        export_title: str = "",
```
Having this will make the typing easier downwards too.

suricactus commented on libqfieldsync/offliners.py:
```diff
@@ -444,11 +456,15 @@ class LayerInfo(NamedTuple):
-        project_title = project.title()
-        if not project_title:
-            project_title = QFileInfo(project.fileName()).baseName()
+        if exported_project_title:
+            project_title = exported_project_title
+        else:
+            project_title = project.title()
+            if not project_title:
+                project_title = QFileInfo(project.fileName()).baseName()
+            project_title += f"{project_title} (offline)"
+        project.setTitle(project_title)
```
```suggestion
        if exported_project_title:
            project_title = exported_project_title
        else:
            project_title = project.title()

            if not project_title:
                project_title = QFileInfo(project.fileName()).baseName()

            project_title += f"{project_title} (offline)"

        project.setTitle(project_title)
```
Make it breathe :)

suricactus commented on libqfieldsync/utils/qgis.py:
```diff
@@ -48,7 +48,9 @@ def open_project(filename: str, filename_to_read: Optional[str] = None) -> bool:
 
 
-def make_temp_qgis_file(project: QgsProject) -> str:
+def make_temp_qgis_file(
+    project: QgsProject,
+) -> str:
```
```suggestion
def make_temp_qgis_file(project: QgsProject) -> str:
```
No need for this diff at all.

suricactus reviewed (changes requested):
Ok, now it's mostly styling stuff, all the rest looks good !

SeqLaz commented:
Hey @suricactus, in this last commit, I addressed all the suggestions. Have a look!

SeqLaz commented:
> Ok, now it's mostly styling stuff, all the rest looks good !

Hey @suricactus, in this last commit, I addressed all the suggestions. Have a look!

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -266,15 +268,15 @@ def _convert(self, project: QgsProject) -> None:
             ):
-                copied_files = layer_source.copy(self.export_folder, copied_files)
+                copied_files = layer_source.copy(
+                    self._export_filename.parent, copied_files
```
```suggestion
                    self._export_filename.parent,
                    copied_files,
```

It is all parameters to be on separate lines when they cannot fit on the same line as the parantesis.

suricactus replied:
```suggestion
                    self._export_filename.parent,
                    copied_files,
```

It is all parameters to be on separate lines when they cannot fit on the same line as the parantesis.

suricactus commented on libqfieldsync/offliners.py:
```diff
@@ -53,6 +53,7 @@ def convert_to_offline(
         offline_db_filename: str,
         layers: List[QgsMapLayer],
         bbox: Optional[QgsRectangle],
+        exported_project_title: Optional[str] = "",
```
```suggestion
        exported_project_title: str = "",
```

This is not optional anymore.

suricactus commented on libqfieldsync/offliners.py:
```diff
@@ -86,6 +87,7 @@ def convert_to_offline(
         offline_db_filename: str,
         layers: List[QgsMapLayer],
         bbox: Optional[QgsRectangle],
+        exported_project_title: Optional[str] = "",
```
```suggestion
        exported_project_title: str = "",
```

This is not optional anymore.

suricactus commented on libqfieldsync/offliners.py:
```diff
@@ -144,8 +149,14 @@ def convert_to_offline(
         offline_db_filename: str,
         layers: List[QgsMapLayer],
         bbox: Optional[QgsRectangle],
+        exported_project_title: Optional[str] = "",
```
```suggestion
        exported_project_title: str = "",
```

This is not optional anymore.

suricactus commented on libqfieldsync/offliners.py:
```diff
@@ -351,6 +362,7 @@ def _convert_to_offline_project(
         offline_gpkg_path: str,
         offline_layers: Optional[List[QgsMapLayer]],
         bbox: Optional[QgsRectangle],
+        exported_project_title: Optional[str] = "",
```
```suggestion
        exported_project_title: str = "",
```

This is not optional anymore.

suricactus commented on libqfieldsync/utils/qgis.py:
```diff
@@ -19,8 +19,8 @@
 """
 
-import logging
 import os
+import logging
 import tempfile
```
Why is that? This should not happen, because the imports are sorted alphabetically, do you have `pre-commit` installed?

SeqLaz replied:
Yes, I have it, and it did not suggest that to me, but from now on, I will take that into consideration. Thanks!

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -102,7 +103,8 @@ def __init__(
         self.trUtf8 = self.tr
 
-        self.export_folder = Path(export_folder)
+        self._export_filename = Path(export_filename)
```
```suggestion
        if not export_filename:
        	raise Exception("Empty export filename provided!")

        self._export_filename = Path(export_filename)
```

We need to be sure it is an existing path.

suricactus reviewed (changes requested):
Thanks, a new round of review.

SeqLaz commented:
Hey @suricactus I added the suggestion, have a look!

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -543,7 +550,7 @@ def _export_basemap(self) -> bool:
             "MAP_UNITS_PER_PIXEL": self.project_configuration.base_map_mupp,
             "MAKE_BACKGROUND_TRANSPARENT": False,
-            "OUTPUT": os.path.join(self.export_folder, "basemap.gpkg"),
+            "OUTPUT": os.path.join(self._export_filename.parent, "basemap.gpkg"),
```
```suggestion
            "OUTPUT": str(self._export_filename.with_name("basemap.gpkg")),
```

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -306,7 +312,7 @@ def _convert(self, project: QgsProject) -> None:
             )
 
-        gpkg_filename = str(self.export_folder.joinpath("data.gpkg"))
+        gpkg_filename = str(self._export_filename.parent.joinpath("data.gpkg"))
```
You can kill this.

suricactus commented on libqfieldsync/offline_converter.py:
```diff
@@ -320,6 +326,7 @@ def _convert(self, project: QgsProject) -> None:
                 gpkg_filename,
```
and add it here:
```
str(self._export_filename.with_name("data.gpkg"))
```

suricactus reviewed (changes requested):
I think this is the last one.

SeqLaz commented:
Hey @suricactus, I've made the suggested changes for the basemap.gpkg and data.gpkg, and also fixed the tests.

Thanks for the suggestions!

suricactus reviewed (approved):
Looks good! I will merge it when the upstream is adapted.
